### PR TITLE
Typescript uno/othermetas/hangman

### PR DESCRIPTION
**Pull request by KrisXV**

### CODE REVIEW

File: server/chat-plugins/uno.ts [hunk 1]
```diff
@@ -216,29 +206,21 @@ class UnoGame extends Rooms.RoomGame {
 
 	/**
 	 * Overwrite the default makePlayer so it makes an UnoGamePlayer instead.
-	 * @param {User} user
-	 * @return {UnoGamePlayer}
 	 */
-	makePlayer(user) {
+	makePlayer(user: User): UnoGamePlayer {
```
Comment (whalemer) [inline server/chat-plugins/uno.ts]: Generally we don't annotate return types unless typescript can't figure it out itself.

File: server/chat-plugins/uno.ts [hunk 2]
```diff
@@ -286,46 +264,39 @@ class UnoGame extends Rooms.RoomGame {
 		return name;
 	}
 
-	/**
-	 * @param {string} msg
-	 * @param {boolean} [overrideSuppress]
-	 */
-	sendToRoom(msg, overrideSuppress = false) {
+	sendToRoom(msg: string, overrideSuppress: boolean = false) {
```
Comment (TheJetOU) [inline server/chat-plugins/uno.ts]:
```suggestion
	sendToRoom(msg: string, overrideSuppress = false) {
```

This is inferrable by the compiler.

File: server/chat-plugins/uno.ts [hunk 3]
```diff
@@ -365,53 +333,41 @@ class UnoGame extends Rooms.RoomGame {
...
-		let userList = Object.keys(this.playerTable);
+		const userList = Object.keys(this.playerTable);
 
 		let player = userList[(userList.indexOf(this.currentPlayerid) + this.direction)];
 
 		if (!player) {
 			player = this.direction === 1 ? userList[0] : userList[this.playerCount - 1];
 		}
-		return player;
+		return player as ID;
```
Comment (TheJetOU) [inline server/chat-plugins/uno.ts]: Why not just define `player` as `let player: ID` for extra type safeness ?

File: server/chat-plugins/uno.ts [hunk 4]
```diff
@@ -435,7 +401,7 @@ class UnoGame extends Rooms.RoomGame {
 		// update the unoId here, so when the display is sent to the player when the play is made
 		if (player.hand.length === 1) {
 			this.awaitUno = player.id;
-			this.unoId = Math.floor(Math.random() * 100).toString();
+			this.unoId = toID(Math.floor(Math.random() * 100).toString());
```
Comment (TheJetOU) [inline server/chat-plugins/uno.ts]:
```suggestion
			this.unoId = Math.floor(Math.random() * 100).toString() as ID;
```

File: server/chat-plugins/uno.ts [hunk 5]
```diff
@@ -526,33 +490,28 @@ class UnoGame extends Rooms.RoomGame {
...
-	 * @param {UnoGamePlayer} player
-	 * @param {number} count
-	 * @return {Card[]}
-	 */
-	onDrawCard(player, count) {
+	onDrawCard(player: UnoGamePlayer, count: number) {
 		if (typeof count === 'string') count = parseInt(count);
 		if (!count || isNaN(count) || count < 1) count = 1;
-		let drawnCards = /** @type {Card[]} */ (this.drawCard(count));
+		const drawnCards: Card[] = this.drawCard(count);
```
Comment (TheJetOU) [inline server/chat-plugins/uno.ts]:
```suggestion
		const drawnCards = this.drawCard(count);
```

I think this is inferrable too.

File: server/chat-plugins/uno.ts [hunk 6]
```diff
@@ -561,12 +520,7 @@ class UnoGame extends Rooms.RoomGame {
 		return drawnCards;
 	}
 
-	/**
-	 * @param {UnoGamePlayer} player
-	 * @param {string} unoId
-	 * @return {false | void}
-	 */
-	onUno(player, unoId) {
+	onUno(player: UnoGamePlayer, unoId: string) {
```
Comment (TheJetOU) [inline server/chat-plugins/uno.ts]:
```suggestion
	onUno(player: UnoGamePlayer, unoId: ID) {
```

File: server/chat-plugins/uno.ts (hunk 4 above)
Comment (KrisXV) [reply server/chat-plugins/uno.ts]: im not sure what difference this makes

File: server/chat-plugins/uno.ts (hunk 2 above)
Comment (KrisXV) [reply server/chat-plugins/uno.ts]: i know but id prefer to keep it defined in the case that something is added later on that introduces potential complexities

File: server/chat-plugins/uno.ts (hunk 4 above)
Comment (TheJetOU) [reply server/chat-plugins/uno.ts]: You're changing code that's used in run-time to appease TypeScript, which only runs at compile-time. There's no reason to do that unless it's the only solution.

File: server/rooms.ts [hunk 7]
```diff
@@ -157,6 +158,7 @@ export abstract class BasicRoom {
 		this.unoDisabled = false;
 		this.blackjackDisabled = false;
 		this.hangmanDisabled = false;
+		this.giveaways = Object.create(null);
```
Comment (Zarel) [inline server/rooms.ts]: ??? no!

Comment (Zarel) [reply server/rooms.ts]:
```suggestion
		this.giveaways = null;
```

File: server/rooms.ts [hunk 8]
```diff
@@ -101,6 +101,7 @@ export abstract class BasicRoom {
 	unoDisabled: boolean;
 	blackjackDisabled: boolean;
 	hangmanDisabled: boolean;
+	giveaways: {wifi: any, gts: any};
```
Comment (Zarel) [inline server/rooms.ts]:
```suggestion
	giveaways: {wifi: any, gts: any} | null;
```

File: server/rooms.ts (hunk 7 above)
Comment (Zarel) [reply server/rooms.ts]: Actually I'm not sure this should be in the main room object at all.

**Comment by Zarel** [comment]
Going to just merge because why not

**Pull request merged**